runtime/addrmap.c

```diff
@@ -0,0 +1,134 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*                 Stephen Dolan, University of Cambridge                 */
+/*                                                                        */
+/*   Copyright 2015 University of Cambridge                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#include "caml/config.h"
+#include "caml/memory.h"
+#include "caml/addrmap.h"
+
+#define MAX_CHAIN 100
+
+static uintnat pos_initial(struct addrmap* t, value key)
+{
+  uintnat pos = (uintnat)key;
+  pos *= 0xcc9e2d51;
+  pos ^= (pos >> 17);
+
+  CAMLassert(Is_power_of_2(t->size));
+  return pos & (t->size - 1);
+}
+
+static uintnat pos_next(struct addrmap* t, uintnat pos)
+{
+  return (pos + 1) & (t->size - 1);
+}
+
+int caml_addrmap_contains(struct addrmap* t, value key)
+{
+  uintnat pos, i;
+
+  CAMLassert(Is_block(key));
+  if (!t->entries) return 0;
+
+  for (i = 0, pos = pos_initial(t, key);
+       i < MAX_CHAIN;
+       i++, pos = pos_next(t, pos)) {
+    if (t->entries[pos].key == ADDRMAP_INVALID_KEY) break;
+    if (t->entries[pos].key == key) return 1;
+  }
+  return 0;
+}
+
+value caml_addrmap_lookup(struct addrmap* t, value key)
+{
+  uintnat pos;
+
+  CAMLassert(Is_block(key));
+  CAMLassert(t->entries);
+
+  for (pos = pos_initial(t, key); ; pos = pos_next(t, pos)) {
+    CAMLassert(t->entries[pos].key != ADDRMAP_INVALID_KEY);
+    if (t->entries[pos].key == key)
+      return t->entries[pos].value;
+  }
+}
+
+static void addrmap_alloc(struct addrmap* t, uintnat sz)
+{
+  uintnat i;
+  CAMLassert(sz > 0 && (sz & (sz - 1)) == 0); /* sz must be a power of 2 */
```

Comment:
Line 27 uses the `Is_power_of_2` macro for this purpose.

Reply:
Fixed in e74f2d9